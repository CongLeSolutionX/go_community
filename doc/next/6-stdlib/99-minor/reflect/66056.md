The new method add a similar for range capability.
[Type.CanSeq]/[Type.CanSeq2] report whether the type is convertible to [iter.Seq]/[iter.Seq2].
[Value.Seq] and [Value.Seq2] Similar to reflection for range of a value.