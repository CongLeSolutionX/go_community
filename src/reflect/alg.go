// Copyright 2015 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package reflect

import "unsafe"

// type algorithms - known to compiler
// a copy of runtime/alg.go - keep in sync
const (
	alg_MEM = iota
	alg_MEM0
	alg_MEM8
	alg_MEM16
	alg_MEM32
	alg_MEM64
	alg_MEM128
	alg_NOEQ
	alg_NOEQ0
	alg_NOEQ8
	alg_NOEQ16
	alg_NOEQ32
	alg_NOEQ64
	alg_NOEQ128
	alg_STRING
	alg_INTER
	alg_NILINTER
	alg_SLICE
	alg_FLOAT32
	alg_FLOAT64
	alg_CPLX64
	alg_CPLX128
	alg_max
)

// a copy of runtime.typeAlg
type typeAlg struct {
	// function for hashing objects of this type
	// (ptr to object, size, seed) -> hash
	hash func(unsafe.Pointer, uintptr, uintptr) uintptr
	// function for comparing objects of this type
	// (ptr to object A, ptr to object B, size) -> ==?
	equal func(unsafe.Pointer, unsafe.Pointer, uintptr) bool
}

// implemented in runtime as reflect_algarray
func algarray(i int) *typeAlg

// algtype1 returns the index into algarray or -1 and the associated typeAlg (or nil)
//
// algtype1 is modeled after cmd/gc/subr.c's algtype1
func algtype1(t Type) (int, *typeAlg) {
	switch t.Kind() {
	case Int8, Int16, Int32, Int64,
		Uint8, Uint16, Uint32, Uint64,
		Int, Uint, Uintptr,
		Bool,
		Ptr,
		Chan, UnsafePointer:
		return alg_MEM, algarray(alg_MEM)

	case Func, Map:
		return alg_NOEQ, algarray(alg_NOEQ)

	case Float32:
		return alg_FLOAT32, algarray(alg_FLOAT32)

	case Float64:
		return alg_FLOAT64, algarray(alg_FLOAT64)

	case Complex64:
		return alg_CPLX64, algarray(alg_CPLX64)

	case Complex128:
		return alg_CPLX128, algarray(alg_CPLX128)

	case String:
		return alg_STRING, algarray(alg_STRING)

	case Interface:
		if t.NumMethod() <= 0 {
			return alg_NILINTER, algarray(alg_NILINTER)
		}
		return alg_INTER, algarray(alg_INTER)

	case Array:
		et := t.Elem()
		alg := et.common().alg
		if alg == nil {
			panic("reflect: type " + et.Name() + " has nil typeAlg")
		}
		return -1, alg

	case Slice:
		return alg_SLICE, algarray(alg_SLICE)

	case Struct:
		panic("not implemented")

	default:
		panic("unknown kind: " + t.Kind().String())
	}

	return 0, nil
}

// algtype returns the correct typeAlg for the given Type.
//
// algtype is modeled after cmd/gc/subr.c's algtype
func algtype(t Type) *typeAlg {
	i, alg := algtype1(t)
	if i == alg_MEM || i == alg_NOEQ {
		switch t.Size() {
		case 0:
			i += alg_MEM0 - alg_MEM
		case 1:
			i += alg_MEM8 - alg_MEM
		case 2:
			i += alg_MEM16 - alg_MEM
		case 4:
			i += alg_MEM32 - alg_MEM
		case 8:
			i += alg_MEM64 - alg_MEM
		case 16:
			i += alg_MEM128 - alg_MEM
		}
		alg = algarray(i)
	}

	if alg == nil {
		panic("reflect: nil typeAlg for type " + t.Name())
	}

	return alg
}
