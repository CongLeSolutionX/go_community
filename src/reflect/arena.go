// Copyright 2017 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package reflect

import (
	"unsafe"
)

type Arena struct {
	ptr  unsafe.Pointer // allocation point
	size uintptr        // remaining size
}

type arenaStore struct {
	// This type must contain at least one pointer so the arena
	// allocation doesn't happen in a noscan span.
	// As soon as we allocate the arena, we will erase its pointer bit.
	_ unsafe.Pointer
}

var (
	arenaStoreIface = interface{}(arenaStore{})
	arenaStoreType  = (*emptyInterface)(unsafe.Pointer(&arenaStoreIface)).typ
	uintptrIface    = interface{}(uintptr(0))
	uintptrType     = (*emptyInterface)(unsafe.Pointer(&uintptrIface)).typ
)

func NewArena(bytes int) *Arena {
	if bytes <= 0 {
		return new(Arena)
	}
	b := uintptr(bytes)
	if b < unsafe.Sizeof(arenaStore{}) {
		b = unsafe.Sizeof(arenaStore{})
	}

	// TODO: Might as well round up to size class.

	typ := *arenaStoreType
	typ.size = b
	base := unsafe_New(&typ)
	// Overwrite the single ptr bit.
	// We need to clobber the pointer bit in case the entire arena is
	// filled with pointerless allocations.
	unsafe_NewAt(uintptrType, base)

	// TODO: Arenas are allocated in the heap. Return them by value instead?
	// Then we'd have to return updated arenas on every operation.
	return &Arena{ptr: base, size: b}
}

var zeroBase byte

func (a *Arena) reservePointer(size uintptr, align uint8) unsafe.Pointer {
	// TODO: Atomic to make this thread-safe. (Otherwise this is
	// an easy way to escape memory safety, even though
	// technically we could call it a race.)
	p := a.ptr
	s := a.size

	// Allocate from the beginning.
	pad := (-uintptr(p)) & (uintptr(align) - 1) // alignment padding
	size += pad
	if size > s {
		return nil
	}
	a.size = s - size
	if a.size != 0 {
		a.ptr = add(p, size, "remaining size > 0")
	} else {
		// Prevent pointer-to-next-object-in-heap.
		a.ptr = nil
	}
	return add(p, pad, "size > 0")
}

func (a *Arena) reserveScalar(size uintptr, align uint8) unsafe.Pointer {
	if size == 0 {
		return unsafe.Pointer(&zeroBase)
	}
	p := a.ptr
	s := a.size
	// Allocate from the end.
	pad := (uintptr(p) + s) & (uintptr(align) - 1) // alignment padding
	size += pad
	if size > s {
		return nil
	}
	a.size = s - size
	return add(p, s-size, "size > 0")
}

// New allocates a value of type typ and returns a pointer to it.
// If there isn't enough space left in the arena, returns nil.
func (a *Arena) New(typ Type) interface{} {
	if typ == nil {
		panic("reflect: Arena.New(nil)")
	}
	rt := typ.(*rtype)
	var ptr unsafe.Pointer
	if rt.pointers() {
		ptr = a.reservePointer(rt.size, rt.align)
		if ptr != nil {
			unsafe_NewAt(rt, ptr)
		}
	} else {
		ptr = a.reserveScalar(rt.size, rt.align)
	}
	if ptr == nil {
		return nil
	}
	i := emptyInterface{
		typ:  rt.ptrTo(),
		word: ptr,
	}
	return *(*interface{})(unsafe.Pointer(&i))
}

// res must be *[]T.  Allocates an array a = [cap]T, assigns *res = a[:].
// The weird argument convention is so the storage for the slice header
// can be allocated in the caller's stack frame.
// If the array won't fit in the arena, returns without allocating or assigning anything.
func (a *Arena) Slice(cap int, res interface{}) {
	if cap < 0 {
		panic("reflect.Arena.MakeSlice: negative cap")
	}
	i := (*emptyInterface)(unsafe.Pointer(&res))
	t := i.typ
	if t.Kind() != Ptr {
		panic("reflect.Arena.Slice result of non-ptr type")
	}
	t = (*ptrType)(unsafe.Pointer(t)).elem
	if t.Kind() != Slice {
		panic("reflect.Arena.Slice of non-ptr-to-slice type")
	}
	t = (*sliceType)(unsafe.Pointer(t)).elem
	// t is now the element type of the slice we want to allocate.

	size := t.size * uintptr(cap) // TODO: Check for overflow

	var ptr unsafe.Pointer
	if t.pointers() {
		ptr = a.reservePointer(size, t.align)
		if ptr != nil {
			unsafe_NewArrayAt(t, cap, ptr)
		}
	} else {
		ptr = a.reserveScalar(size, t.align)
	}
	if ptr != nil {
		*(*sliceHeader)(i.word) = sliceHeader{Data: ptr, Len: cap, Cap: cap}
	}
}

// implemented in package runtime
func unsafe_NewAt(*rtype, unsafe.Pointer)
func unsafe_NewArrayAt(*rtype, int, unsafe.Pointer)
