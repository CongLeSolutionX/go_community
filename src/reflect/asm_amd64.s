// Copyright 2012 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#include "textflag.h"
#include "funcdata.h"
#include "go_asm.h"

// makeFuncStub is the code half of the function returned by MakeFunc.
// See the comment on the declaration of makeFuncStub in makefunc.go
// for more details.
// No arg size here; runtime pulls arg map out of the func value.
// makeFuncStub must be ABIInternal because it is placed directly
// in function values.
// The frame contains a stack-allocated RegArgs after the space saved for
// arguments to callReflect. This RegArgs is visible to the GC via a special
// case in the runtime's stack map code, so its offset is known to the runtime.
//
// TODO(mknyszek): Modify this function to call callReflect with
// ABIInternal.
TEXT ·makeFuncStub<ABIInternal>(SB),(NOSPLIT|WRAPPER),$312
	NO_LOCAL_POINTERS
	LEAQ	40(SP), R12
	CALL	runtime·spillArgs<ABIInternal>(SB)
	MOVQ	DX, 0(SP)
	MOVQ	R12, 8(SP)
	CALL	·moveMakeFuncArgPtrs(SB)
	LEAQ	argframe+0(FP), CX
	MOVQ	CX, 8(SP)
	MOVB	$0, 32(SP)
	LEAQ	32(SP), AX
	MOVQ	AX, 16(SP)
	LEAQ	40(SP), AX
	MOVQ	AX, 24(SP)
	CALL	·callReflect<ABIInternal>(SB)
	LEAQ	40(SP), R12
	CALL	runtime·unspillArgs<ABIInternal>(SB)
	RET

// methodValueCall is the code half of the function returned by makeMethodValue.
// See the comment on the declaration of methodValueCall in makefunc.go
// for more details.
// No arg size here; runtime pulls arg map out of the func value.
// methodValueCall must be ABIInternal because it is placed directly
// in function values.
// The frame contains a stack-allocated RegArgs after the space saved for
// arguments to callMethod. This RegArgs is visible to the GC via a special
// case in the runtime's stack map code, so its offset is known to the runtime.
//
// TODO(mknyszek): Modify this function to call callMethod with
// ABIInternal.
TEXT ·methodValueCall<ABIInternal>(SB),(NOSPLIT|WRAPPER),$312
	NO_LOCAL_POINTERS
	LEAQ	40(SP), R12
	CALL	runtime·spillArgs<ABIInternal>(SB)
	MOVQ	DX, 0(SP)
	MOVQ	R12, 8(SP)
	CALL	·moveMakeFuncArgPtrs(SB)
	LEAQ	argframe+0(FP), CX
	MOVQ	CX, 8(SP)
	MOVB	$0, 32(SP)
	LEAQ	32(SP), AX
	MOVQ	AX, 16(SP)
	LEAQ	40(SP), AX
	MOVQ	AX, 24(SP)
	CALL	·callMethod<ABIInternal>(SB)
	LEAQ	40(SP), R12
	CALL	runtime·unspillArgs<ABIInternal>(SB)
	RET
