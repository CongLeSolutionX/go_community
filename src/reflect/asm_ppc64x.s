// Copyright 2012 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

//go:build ppc64 || ppc64le
// +build ppc64 ppc64le

#include "textflag.h"
#include "funcdata.h"
#include "asm_ppc64x.h"

// FIXED_FRAME+32 is a save spot
// FIXED_FRAME+40 is return information
// FIXED_FRAME+48 is start of register args

#define SAVER11 FIXED_FRAME
#define LOCAL_RETVALID 40+FIXED_FRAME
#define LOCAL_REGARGS 48+FIXED_FRAME

// The frame size of the functions below is
// 32 fixed frame + 32 (args of callReflect) + 8 (bool + padding) + 288 (abi.RegArgs) = 360.

// makeFuncStub is the code half of the function returned by MakeFunc.
// See the comment on the declaration of makeFuncStub in makefunc.go
// for more details.
// No arg size here, runtime pulls arg map out of the func value.
TEXT ·makeFuncStub(SB),(NOSPLIT|WRAPPER),$360
	NO_LOCAL_POINTERS
	ADD	$LOCAL_REGARGS, R1, R20
	CALL	runtime·spillArgs(SB)
	MOVD	R11, FIXED_FRAME+32(R1)	// TODO: determine offset to save
	MOVD	R11, FIXED_FRAME+0(R1)	// arg for moveMakeFuncArgPtrs
	MOVD	R20, FIXED_FRAME+8(R1)	// arg for local args
	CALL	·moveMakeFuncArgPtrs(SB)
	MOVD	FIXED_FRAME+32(R1), R11	// restore R11 ctxt
	// previous
	MOVD	R11, FIXED_FRAME+0(R1)	// ctxt (arg0)
	MOVD	$argframe+0(FP), R3	// ??
	MOVD	R3, FIXED_FRAME+8(R1)	// frame (arg1)
	ADD	$LOCAL_RETVALID, R1, R3 // addr of return flag
	MOVB	R0, (R3)		// clear flag
	MOVD	R3, FIXED_FRAME+16(R1)	// addr retvalid (arg2)
	ADD     $LOCAL_REGARGS, R1, R3
	MOVD	R3, FIXED_FRAME+24(R1)	// abiregargs (arg3)
	BL	·callReflect(SB)
	ADD	$LOCAL_REGARGS, R1, R20	// set address of spill area
	CALL	runtime·unspillArgs(SB)
	RET

// methodValueCall is the code half of the function returned by makeMethodValue.
// See the comment on the declaration of methodValueCall in makefunc.go
// for more details.
// No arg size here; runtime pulls arg map out of the func value.
TEXT ·methodValueCall(SB),(NOSPLIT|WRAPPER),$360
	NO_LOCAL_POINTERS
	ADD	$LOCAL_REGARGS, R1, R20
	CALL	runtime·spillArgs(SB)
	MOVD	R11, FIXED_FRAME+0(R1) // arg0 ctxt
	MOVD	R11, FIXED_FRAME+32(R1) // save for later
	MOVD	R20, FIXED_FRAME+8(R1) // arg1 abiregargs
	CALL	·moveMakeFuncArgPtrs(SB)
	MOVD	FIXED_FRAME+32(R1), R11 // restore ctxt
	MOVD	R11, FIXED_FRAME+0(R1) // set as arg0
	MOVD	$argframe+0(FP), R3	// frame pointer
	MOVD	R3, FIXED_FRAME+8(R1)	// set as arg1
	ADD	$LOCAL_RETVALID, R1, R3
	MOVB	$0, (R3)		// clear ret flag
	MOVD	R3, FIXED_FRAME+16(R1)	// addr of return flag
	ADD	$LOCAL_REGARGS, R1, R3	// addr of abiregargs
	MOVD	R3, FIXED_FRAME+24(R1)	// set as arg3
	BL	·callMethod(SB)
	ADD     $LOCAL_REGARGS, R1, R20
	CALL	runtime·unspillArgs(SB)
	RET
