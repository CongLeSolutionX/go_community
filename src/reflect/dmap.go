// Copyright 2016 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package reflect

import (
	"sync"
	"sync/atomic"
)

// A dmap is an distributed atomic map.
// It supports only insert and lookup; there is no delete.
type dmap struct {
	mu      sync.Mutex   // held when writing to root
	root    atomic.Value // *node
	m       atomic.Value // map[interface{}]interface{}
	compare func(interface{}, interface{}) int
}

// A node is a node in the distributed map.
type node struct {
	key, val    interface{}  // never changed after creation
	mu          sync.Mutex   // held when writing to left or right
	left, right atomic.Value // *node
	count       uint32       // number of lookups
}

// New returns a new distribute map given a comparison function.
func newdmap(compare func(interface{}, interface{}) int) *dmap {
	return &dmap{
		compare: compare,
	}
}

// Insert inserts a key/value pair into a dmap.
func (d *dmap) insert(key, val interface{}) {
	var n *node
	for {
		root, _ := d.root.Load().(*node)
		if root != nil {
			n = root
			break
		}
		root = &node{
			key: key,
			val: val,
		}
		d.mu.Lock()
		if d.root.Load() == nil {
			d.root.Store(root)
			d.mu.Unlock()
			return
		}
		d.mu.Unlock()
	}

	for {
		cmp := d.compare(key, n.key)
		if cmp == 0 {
			if val != n.val {
				panic("invalid double-insert")
			}
			return
		}
		p := &n.left
		if cmp > 0 {
			p = &n.right
		}
		n2, _ := (*p).Load().(*node)
		if n2 != nil {
			n = n2
		} else {
			n2 = &node{
				key: key,
				val: val,
			}
			n.mu.Lock()
			if (*p).Load() == nil {
				(*p).Store(n2)
				n.mu.Unlock()
				return
			}
			n.mu.Unlock()
		}
	}
}

// Lookup looks up a key in the distributed map.
func (d *dmap) lookup(key interface{}) interface{} {
	// Common values are cached in a map held in the root.
	m, _ := d.m.Load().(map[interface{}]interface{})
	if val, ok := m[key]; ok {
		return val
	}

	n, _ := d.root.Load().(*node)
	for n != nil {
		cmp := d.compare(key, n.key)
		if cmp == 0 {
			count := atomic.AddUint32(&n.count, 1)

			// Add this key/val pair to the map in the root,
			// but only if it's worth copying the existing map.
			if count < 0 || (count > 1 && int(count) > len(m)) {
				newm := make(map[interface{}]interface{}, len(m)+1)
				for k, v := range m {
					newm[k] = v
				}
				newm[key] = n.val

				// It's possible that some other
				// goroutine has updated d.m since we
				// loaded it.  That means we did extra
				// work but it's otherwise OK.
				d.m.Store(newm)
			}

			return n.val
		}

		p := &n.left
		if cmp > 0 {
			p = &n.right
		}
		n, _ = (*p).Load().(*node)
	}

	return nil
}
