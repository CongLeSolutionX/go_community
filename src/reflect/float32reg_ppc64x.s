// Copyright 2021 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

//go:build ppc64 || ppc64le
// +build ppc64 ppc64le

#include "textflag.h"

// Convert float32->uint64
TEXT ·archFloat32ToReg(SB),NOSPLIT,$0-16
	FMOVS	val+0(FP), F1
	FMOVD	F1, ret+8(FP)
	RET

// Convert uint64->float32
TEXT ·archFloat32FromReg(SB),NOSPLIT,$0-12
	FMOVD	reg+0(FP), F1
	FMOVS	F1, ret+8(FP)
	RET

