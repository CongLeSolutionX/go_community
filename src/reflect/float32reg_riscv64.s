// Copyright 2022 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#include "textflag.h"

// On riscv64, the float32 becomes a float64
// when loaded in a register, different from
// other platforms. These functions are
// needed to ensure correct conversions on riscv64

// Convert float32->uint64
TEXT ·archFloat32ToReg(SB),NOSPLIT,$0-16
	MOVF	val+0(FP), F1
	MOVD	F1, ret+8(FP)
	RET

// Convert uint64->float32
TEXT ·archFloat32FromReg(SB),NOSPLIT,$0-12
	MOVD	reg+0(FP), F1
	// Normally a float64->float32 conversion
	// would need rounding, but that is not needed
	// here since the uint64 was originally converted
	// from float32, and should be avoided to
	// preserve SNaN values.
	MOVF	F1, ret+8(FP)
	RET

