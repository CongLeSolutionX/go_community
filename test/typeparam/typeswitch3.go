// run -gcflags=-G=3

// Copyright 2021 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package main

import "reflect"

type I interface { foo() }

type myint int

func (myint) foo() {}

type myfloat float64
func (myfloat) foo() {}

type myint32 int32
func (myint32) foo() {}

func f[T I](i I) {
	switch x := i.(type) {
	case T:
		println("T", x)
	case myint:
		println("myint", x)
	default:
		println("other", reflect.ValueOf(x).Int())
	}
}
func main() {
	f[myfloat](myint(6))
	f[myfloat](myfloat(7))
	f[myfloat](myint32(8))
}
