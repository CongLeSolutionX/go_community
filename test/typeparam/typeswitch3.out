myint 6
T +7.000000e+000
other 8
